lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,261 @@
+"use strict";
+const http = require("http");
+const https = require("https");
+const { Writable } = require("stream");
+const zlib = require("zlib");
+
+const ver = process.version.replace("v", "").split(".");
+const majorNodeVersion = Number.parseInt(ver[0]);
+
+function abortRequest(clientRequest) {
+  // clientRequest.destroy breaks the test suite for versions 10 and 12,
+  // hence the version check
+  if (majorNodeVersion > 13) {
+    clientRequest.destroy();
+  } else {
+    clientRequest.abort();
+  }
+  clientRequest.removeAllListeners();
+  clientRequest.on("error", () => {});
+}
+
+module.exports = class Request extends Writable {
+  constructor(url, options) {
+    super();
+    this._currentURL = url;
+    this._options = options;
+    this.headers = this._options.headers;
+    this._ended = false;
+    this._redirectCount = 0;
+    this._requestBodyBuffers = [];
+    this._bufferIndex = 0;
+    this.response = Object.create(null);
+    this._performRequest();
+  }
+
+  abort() {
+    abortRequest(this._currentRequest);
+    this.emit("abort");
+    this.removeAllListeners();
+  }
+
+  write(data, encoding) {
+    if (data.length > 0) {
+      this._requestBodyBuffers.push({ data, encoding });
+      this._currentRequest.write(data, encoding);
+    }
+  }
+
+  end() {
+    this.emit("request", this._currentRequest);
+    this._ended = true;
+    this._currentRequest.end();
+  }
+
+  setHeader(name, value) {
+    this._options.headers[name] = value;
+    this._currentRequest.setHeader(name, value);
+  }
+
+  removeHeader(name) {
+    delete this._options.headers[name];
+    this._currentRequest.removeHeader(name);
+  }
+
+  // Without this method, the test suite will halt
+  toJSON() {
+    const { method, headers } = this._options;
+    return { uri: new URL(this._currentURL), method, headers };
+  }
+
+  _writeNext(error) {
+    if (this._currentRequest) {
+      if (error) {
+        this.emit("error", error);
+      } else if (this._bufferIndex < this._requestBodyBuffers.length) {
+        const buffer = this._requestBodyBuffers[this._bufferIndex++];
+        if (!this._currentRequest.writableEnded) {
+          this._currentRequest.write(
+            buffer.data,
+            buffer.encoding,
+            this._writeNext.bind(this)
+          );
+        }
+      } else if (this._ended) {
+        this._currentRequest.end();
+      }
+    }
+  }
+
+  _performRequest() {
+    const urlOptions = new URL(this._currentURL);
+    const scheme = urlOptions.protocol;
+    this._options.agent = this._options.agents[scheme.substring(0, scheme.length - 1)];
```

Comment:
I'm confused by this line. Nothing else in the PR uses `_options.agent` from what I can tell. So I guess this is used on the next line? But now it seems like `this._options` is serving multiple purposes: some of its properties (like `agents`) are inputs to this file, and some of the properties (like `agent`) are input to `http.request()`.

Is it possible to separate these uses out, so that `this._options` only serves a single purpose? Or maybe it goes away entirely, in favor of just underscore-prefixed properties for storing input/state + assembling a new options object in this function.